Modules/_asynciomodule.c

```diff
@@ -1601,11 +1601,24 @@ static void
 FutureIter_dealloc(futureiterobject *it)
 {
     PyTypeObject *tp = Py_TYPE(it);
-    asyncio_state *state = get_asyncio_state_by_def((PyObject *)it);
+
+    assert(_PyType_HasFeature((PyTypeObject *)tp, Py_TPFLAGS_HEAPTYPE));
+
+    PyHeapTypeObject *ht = (PyHeapTypeObject*)tp;
+    PyObject *module = ht->ht_module;
+    asyncio_state *state = NULL;
+
     PyObject_GC_UnTrack(it);
     tp->tp_clear((PyObject *)it);
 
-    if (state->fi_freelist_len < FI_FREELIST_MAXLEN) {
+    // GH-115874: We can't use PyType_GetModuleByDef here as the type might have
+    // already been cleared. This is also why we must check if ht_module != NULL.
+    // Subclasses also cannot make use of the free list.
```

Comment:
Maybe reword this sentence to better capture the causal relationship here (they can't use the free list *because* of this restriction).

And, more pedantically, it's subclasses that belong to a different module that can't use it.

Reply:
I agree, this line threw me off (I was expecting an "exact" type check on the next line, but no).

The rest LGTM!